ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs

```diff
@@ -0,0 +1,238 @@
+use std::marker::PhantomData;
+
+use ceno_emul::{InsnCategory, InsnCodes, InsnFormat, InsnKind, StepRecord};
+use ff_ext::ExtensionField;
+
+use super::super::{
+    RIVInstruction,
+    constants::UInt,
+    insn_base::{ReadMEM, ReadRS1, ReadRS2, StateInOut, WriteMEM, WriteRD},
+};
+use crate::{
+    circuit_builder::CircuitBuilder,
+    error::ZKVMError,
+    expression::{ToExpr, WitIn},
+    instructions::Instruction,
+    set_val,
+    tables::InsnRecord,
+    uint::Value,
+    utils::i64_to_base,
+    witness::LkMultiplicity,
+};
+use core::mem::MaybeUninit;
+
+/// DummyInstruction can handle any instruction and produce its side-effects.
+pub struct DummyInstruction<E, I>(PhantomData<(E, I)>);
+
+impl<E: ExtensionField, I: RIVInstruction> Instruction<E> for DummyInstruction<E, I> {
+    type InstructionConfig = DummyConfig<E>;
+
+    fn name() -> String {
+        format!("{:?}_DUMMY", I::INST_KIND)
+    }
+
+    fn construct_circuit(
+        circuit_builder: &mut CircuitBuilder<E>,
+    ) -> Result<Self::InstructionConfig, ZKVMError> {
+        DummyConfig::construct_circuit(circuit_builder, I::INST_KIND.codes())
+    }
+
+    fn assign_instance(
+        config: &Self::InstructionConfig,
+        instance: &mut [MaybeUninit<<E as ExtensionField>::BaseField>],
+        lk_multiplicity: &mut LkMultiplicity,
+        step: &StepRecord,
+    ) -> Result<(), ZKVMError> {
+        config.assign_instance(instance, lk_multiplicity, step)
+    }
+}
+
+#[derive(Debug)]
+pub struct DummyConfig<E: ExtensionField> {
+    vm_state: StateInOut<E>,
+
+    rs1: Option<(ReadRS1<E>, UInt<E>)>,
+    rs2: Option<(ReadRS2<E>, UInt<E>)>,
+    rd: Option<(WriteRD<E>, UInt<E>)>,
+
+    mem_addr_val: Option<[WitIn; 3]>,
+    mem_read: Option<ReadMEM<E>>,
+    mem_write: Option<WriteMEM>,
+
+    imm: WitIn,
+}
+
+impl<E: ExtensionField> DummyConfig<E> {
+    fn construct_circuit(
+        circuit_builder: &mut CircuitBuilder<E>,
+        codes: InsnCodes,
+    ) -> Result<Self, ZKVMError> {
+        let (with_rs1, with_rs2, with_rd) = match (codes.format, codes.kind) {
+            (_, InsnKind::EANY) => (true, true, false),
```

Comment:
For `ecall` instructions, I think they need to read the `rd` register to get `syscall_id`.

Reply:
It has to match `VMState::ecall`. That’s not true in this PR, I fixed it in another branch. I can backport it here to be consistent.

Reply:
Better idea: just treat it as an I instruction. I’ll make it so in the guest-example. Done.

Reply:
After reworking this a bit: the ignored ecalls just read the syscall_id and do nothing else. Added comments to explain.

Reply:
I also turned it into a parameter: `Platform.unsafe_ecall_nop`